File: specs/typespec/org/org-users.ts
import {
	type EmailAddress,
	type Password,
	type DomainName,
	type TFACode,
	type LanguageCode,
	type FullName,
	type ValidationError,
	newValidationError,
	validateEmailAddress,
	validateEmployerEmail,
	validatePassword,
	validateDomainName,
	validateTFACode,
	validateFullName,
	ERR_REQUIRED,
} from "../common/common";

// Token types
export type OrgSessionToken = string;
export type OrgTFAToken = string;
export type DNSVerificationToken = string;
export type OrgSignupToken = string;
export type OrgInvitationToken = string;
export type OrgPasswordResetToken = string;

// ============================================
// Signup Flow (DNS-based Domain Verification)
// ============================================

export interface OrgInitSignupRequest {
	email: EmailAddress;
	home_region: string;
}

export function validateOrgInitSignupRequest(
	request: OrgInitSignupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email) {
		errs.push(newValidationError("email", ERR_REQUIRED));
	} else {
		// Use employer email validation which blocks personal email domains
		const emailErr = validateEmployerEmail(request.email);
		if (emailErr) {
			errs.push(newValidationError("email", emailErr));
		}
	}

	if (!request.home_region) {
		errs.push(newValidationError("home_region", ERR_REQUIRED));
	}

	return errs;
}

export interface OrgInitSignupResponse {
	domain: DomainName;
	dns_record_name: string;
	token_expires_at: string;
	message: string;
}

export interface OrgGetSignupDetailsRequest {
	signup_token: OrgSignupToken;
}

export function validateOrgGetSignupDetailsRequest(
	request: OrgGetSignupDetailsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.signup_token) {
		errs.push(newValidationError("signup_token", ERR_REQUIRED));
	}

	return errs;
}

export interface OrgGetSignupDetailsResponse {
	domain: DomainName;
}

export interface OrgCompleteSignupRequest {
	signup_token: OrgSignupToken;
	password: Password;
	preferred_language: LanguageCode;
	has_added_dns_record: boolean;
	agrees_to_eula: boolean;
}

const ERR_DNS_RECORD_NOT_CONFIRMED =
	"You must confirm that you have added the DNS record";
const ERR_EULA_NOT_ACCEPTED =
	"You must agree to the End User License Agreement";

export function validateOrgCompleteSignupRequest(
	request: OrgCompleteSignupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.signup_token) {
		errs.push(newValidationError("signup_token", ERR_REQUIRED));
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	} else {
		const passwordErr = validatePassword(request.password);
		if (passwordErr) {
			errs.push(newValidationError("password", passwordErr));
		}
	}

	if (!request.preferred_language) {
		errs.push(newValidationError("preferred_language", ERR_REQUIRED));
	}

	if (!request.has_added_dns_record) {
		errs.push(
			newValidationError("has_added_dns_record", ERR_DNS_RECORD_NOT_CONFIRMED)
		);
	}

	if (!request.agrees_to_eula) {
		errs.push(newValidationError("agrees_to_eula", ERR_EULA_NOT_ACCEPTED));
	}

	return errs;
}

export interface OrgCompleteSignupResponse {
	session_token: OrgSessionToken;
	org_user_id: string;
}

// ============================================
// Login Flow
// ============================================

export interface OrgLoginRequest {
	email: EmailAddress;
	domain: DomainName;
	password: Password;
}

export function validateOrgLoginRequest(
	request: OrgLoginRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email) {
		errs.push(newValidationError("email", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email);
		if (emailErr) {
			errs.push(newValidationError("email", emailErr));
		}
	}

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	} else {
		const passwordErr = validatePassword(request.password);
		if (passwordErr) {
			errs.push(newValidationError("password", passwordErr));
		}
	}

	return errs;
}

export interface OrgLoginResponse {
	tfa_token: OrgTFAToken;
}

export interface OrgTFARequest {
	tfa_token: OrgTFAToken;
	tfa_code: TFACode;
	remember_me: boolean;
}

export function validateOrgTFARequest(
	request: OrgTFARequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.tfa_token) {
		errs.push(newValidationError("tfa_token", ERR_REQUIRED));
	}

	if (!request.tfa_code) {
		errs.push(newValidationError("tfa_code", ERR_REQUIRED));
	} else {
		const tfaErr = validateTFACode(request.tfa_code);
		if (tfaErr) {
			errs.push(newValidationError("tfa_code", tfaErr));
		}
	}

	return errs;
}

export interface OrgTFAResponse {
	session_token: OrgSessionToken;
	preferred_language: LanguageCode;
	employer_name: string;
}

// OrgLogoutRequest is empty - session token passed via Authorization header
export interface OrgLogoutRequest {}

// ============================================
// User Invitation Flow
// ============================================

export interface OrgInviteUserRequest {
	email_address: EmailAddress;
	full_name: FullName;
}

export function validateOrgInviteUserRequest(
	request: OrgInviteUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email_address);
		if (emailErr) {
			errs.push(newValidationError("email_address", emailErr));
		}
	}

	if (!request.full_name) {
		errs.push(newValidationError("full_name", ERR_REQUIRED));
	} else {
		const fullNameErr = validateFullName(request.full_name);
		if (fullNameErr) {
			errs.push(newValidationError("full_name", fullNameErr));
		}
	}

	return errs;
}

export interface OrgInviteUserResponse {
	invitation_id: string;
	expires_at: string;
}

export interface OrgCompleteSetupRequest {
	invitation_token: OrgInvitationToken;
	password: Password;
	full_name: FullName;
}

export function validateOrgCompleteSetupRequest(
	request: OrgCompleteSetupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.invitation_token) {
		errs.push(newValidationError("invitation_token", ERR_REQUIRED));
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	} else {
		const passwordErr = validatePassword(request.password);
		if (passwordErr) {
			errs.push(newValidationError("password", passwordErr));
		}
	}

	if (!request.full_name) {
		errs.push(newValidationError("full_name", ERR_REQUIRED));
	} else {
		const fullNameErr = validateFullName(request.full_name);
		if (fullNameErr) {
			errs.push(newValidationError("full_name", fullNameErr));
		}
	}

	return errs;
}

export interface OrgCompleteSetupResponse {
	message: string;
}

// ============================================
// User Management (Disable/Enable)
// ============================================

export interface OrgDisableUserRequest {
	target_user_id: string;
}

export function validateOrgDisableUserRequest(
	request: OrgDisableUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.target_user_id) {
		errs.push(newValidationError("target_user_id", ERR_REQUIRED));
	}

	return errs;
}

export interface OrgEnableUserRequest {
	target_user_id: string;
}

export function validateOrgEnableUserRequest(
	request: OrgEnableUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.target_user_id) {
		errs.push(newValidationError("target_user_id", ERR_REQUIRED));
	}

	return errs;
}

// ============================================================================
// Org Password Management
// ============================================================================

export interface OrgRequestPasswordResetRequest {
	email_address: EmailAddress;
	domain: DomainName;
}

export function validateOrgRequestPasswordResetRequest(
	request: OrgRequestPasswordResetRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email_address);
		if (emailErr) {
			errs.push(newValidationError("email_address", emailErr));
		}
	}

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	return errs;
}

export interface OrgRequestPasswordResetResponse {
	message: string;
}

export interface OrgCompletePasswordResetRequest {
	reset_token: OrgPasswordResetToken;
	new_password: Password;
}

export function validateOrgCompletePasswordResetRequest(
	request: OrgCompletePasswordResetRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.reset_token) {
		errs.push(newValidationError("reset_token", ERR_REQUIRED));
	}

	if (!request.new_password) {
		errs.push(newValidationError("new_password", ERR_REQUIRED));
	} else {
		const passwordErr = validatePassword(request.new_password);
		if (passwordErr) {
			errs.push(newValidationError("new_password", passwordErr));
		}
	}

	return errs;
}

export interface OrgChangePasswordRequest {
	current_password: Password;
	new_password: Password;
}

export function validateOrgChangePasswordRequest(
	request: OrgChangePasswordRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.current_password) {
		errs.push(newValidationError("current_password", ERR_REQUIRED));
	} else {
		const currentPasswordErr = validatePassword(request.current_password);
		if (currentPasswordErr) {
			errs.push(newValidationError("current_password", currentPasswordErr));
		}
	}

	if (!request.new_password) {
		errs.push(newValidationError("new_password", ERR_REQUIRED));
	} else {
		const newPasswordErr = validatePassword(request.new_password);
		if (newPasswordErr) {
			errs.push(newValidationError("new_password", newPasswordErr));
		}
	}

	// Check if current and new passwords are the same
	if (
		request.current_password &&
		request.new_password &&
		request.current_password === request.new_password
	) {
		errs.push(
			newValidationError(
				"new_password",
				"New password must be different from current password"
			)
		);
	}

	return errs;
}

File: specs/typespec/org/org-users.go
package org

import (
	"errors"

	"vetchium-api-server.typespec/common"
)

// Token types
type OrgSessionToken string
type OrgTFAToken string
type DNSVerificationToken string
type OrgSignupToken string
type OrgInvitationToken string
type OrgPasswordResetToken string

// RBAC types (re-exported from common)
type RoleName = common.RoleName
type AssignRoleRequest = common.AssignRoleRequest
type RemoveRoleRequest = common.RemoveRoleRequest

// ============================================
// Signup Flow (DNS-based Domain Verification)
// ============================================

type OrgInitSignupRequest struct {
	Email      common.EmailAddress `json:"email"`
	HomeRegion string              `json:"home_region"`
}

func (r OrgInitSignupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Email == "" {
		errs = append(errs, common.NewValidationError("email", common.ErrRequired))
	} else if err := common.ValidateEmployerEmail(r.Email); err != nil {
		// Use employer email validation which blocks personal email domains
		errs = append(errs, common.NewValidationError("email", err))
	}

	if r.HomeRegion == "" {
		errs = append(errs, common.NewValidationError("home_region", common.ErrRequired))
	}

	return errs
}

type OrgInitSignupResponse struct {
	Domain         common.DomainName `json:"domain"`
	DNSRecordName  string            `json:"dns_record_name"`
	TokenExpiresAt string            `json:"token_expires_at"`
	Message        string            `json:"message"`
}

type OrgGetSignupDetailsRequest struct {
	SignupToken OrgSignupToken `json:"signup_token"`
}

func (r OrgGetSignupDetailsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SignupToken == "" {
		errs = append(errs, common.NewValidationError("signup_token", common.ErrRequired))
	}

	return errs
}

type OrgGetSignupDetailsResponse struct {
	Domain common.DomainName `json:"domain"`
}

type OrgCompleteSignupRequest struct {
	SignupToken       OrgSignupToken      `json:"signup_token"`
	Password          common.Password     `json:"password"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
	HasAddedDNSRecord bool                `json:"has_added_dns_record"`
	AgreesToEULA      bool                `json:"agrees_to_eula"`
}

var (
	errDNSRecordNotConfirmed = errors.New("You must confirm that you have added the DNS record")
	errEULANotAccepted       = errors.New("You must agree to the End User License Agreement")
)

func (r OrgCompleteSignupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SignupToken == "" {
		errs = append(errs, common.NewValidationError("signup_token", common.ErrRequired))
	}

	if r.Password == "" {
		errs = append(errs, common.NewValidationError("password", common.ErrRequired))
	} else if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}

	if r.PreferredLanguage == "" {
		errs = append(errs, common.NewValidationError("preferred_language", common.ErrRequired))
	}

	if !r.HasAddedDNSRecord {
		errs = append(errs, common.NewValidationError("has_added_dns_record", errDNSRecordNotConfirmed))
	}

	if !r.AgreesToEULA {
		errs = append(errs, common.NewValidationError("agrees_to_eula", errEULANotAccepted))
	}

	return errs
}

type OrgCompleteSignupResponse struct {
	SessionToken OrgSessionToken `json:"session_token"`
	OrgUserID    string          `json:"org_user_id"`
}

// ============================================
// Login Flow
// ============================================

type OrgLoginRequest struct {
	Email    common.EmailAddress `json:"email"`
	Domain   common.DomainName   `json:"domain"`
	Password common.Password     `json:"password"`
}

func (r OrgLoginRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Email == "" {
		errs = append(errs, common.NewValidationError("email", common.ErrRequired))
	} else if err := r.Email.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email", err))
	}

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	if r.Password == "" {
		errs = append(errs, common.NewValidationError("password", common.ErrRequired))
	} else if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}

	return errs
}

type OrgLoginResponse struct {
	TFAToken OrgTFAToken `json:"tfa_token"`
}

type OrgTFARequest struct {
	TFAToken   OrgTFAToken    `json:"tfa_token"`
	TFACode    common.TFACode `json:"tfa_code"`
	RememberMe bool           `json:"remember_me"`
}

func (r OrgTFARequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.TFAToken == "" {
		errs = append(errs, common.NewValidationError("tfa_token", common.ErrRequired))
	}

	if r.TFACode == "" {
		errs = append(errs, common.NewValidationError("tfa_code", common.ErrRequired))
	} else if err := r.TFACode.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("tfa_code", err))
	}

	return errs
}

type OrgTFAResponse struct {
	SessionToken      OrgSessionToken     `json:"session_token"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
	EmployerName      string              `json:"employer_name"`
}

// OrgLogoutRequest is empty - session token passed via Authorization header
type OrgLogoutRequest struct{}

// ============================================
// User Invitation Flow
// ============================================

type OrgInviteUserRequest struct {
	EmailAddress      common.EmailAddress `json:"email_address"`
	FullName          common.FullName     `json:"full_name"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
}

func (r OrgInviteUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", common.ErrRequired))
	} else if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	if r.FullName == "" {
		errs = append(errs, common.NewValidationError("full_name", common.ErrRequired))
	} else if err := r.FullName.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("full_name", err))
	}
	if r.PreferredLanguage != "" {
		if err := r.PreferredLanguage.Validate(); err != nil {
			errs = append(errs, common.NewValidationError("preferred_language", err))
		}
	}

	return errs
}

type OrgInviteUserResponse struct {
	InvitationID string `json:"invitation_id"`
	ExpiresAt    string `json:"expires_at"`
}

type OrgCompleteSetupRequest struct {
	InvitationToken OrgInvitationToken `json:"invitation_token"`
	Password        common.Password    `json:"password"`
	FullName        common.FullName    `json:"full_name"`
}

func (r OrgCompleteSetupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.InvitationToken == "" {
		errs = append(errs, common.NewValidationError("invitation_token", common.ErrRequired))
	}

	if r.Password == "" {
		errs = append(errs, common.NewValidationError("password", common.ErrRequired))
	} else if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}

	if r.FullName == "" {
		errs = append(errs, common.NewValidationError("full_name", common.ErrRequired))
	} else if err := r.FullName.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("full_name", err))
	}

	return errs
}

type OrgCompleteSetupResponse struct {
	Message string `json:"message"`
}

// ============================================
// User Management (Disable/Enable)
// ============================================

type OrgDisableUserRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
}

func (r OrgDisableUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	return errs
}

type OrgEnableUserRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
}

func (r OrgEnableUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	return errs
}

// ============================================================================
// Org Password Management
// ============================================================================

type OrgRequestPasswordResetRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
	Domain       common.DomainName   `json:"domain"`
}

func (r OrgRequestPasswordResetRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", common.ErrRequired))
	} else if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type OrgRequestPasswordResetResponse struct {
	Message string `json:"message"`
}

type OrgCompletePasswordResetRequest struct {
	ResetToken  OrgPasswordResetToken `json:"reset_token"`
	NewPassword common.Password       `json:"new_password"`
}

func (r OrgCompletePasswordResetRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.ResetToken == "" {
		errs = append(errs, common.NewValidationError("reset_token", common.ErrRequired))
	}

	if r.NewPassword == "" {
		errs = append(errs, common.NewValidationError("new_password", common.ErrRequired))
	} else if err := r.NewPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("new_password", err))
	}

	return errs
}

type OrgChangePasswordRequest struct {
	CurrentPassword common.Password `json:"current_password"`
	NewPassword     common.Password `json:"new_password"`
}

func (r OrgChangePasswordRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.CurrentPassword == "" {
		errs = append(errs, common.NewValidationError("current_password", common.ErrRequired))
	} else if err := r.CurrentPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("current_password", err))
	}

	if r.NewPassword == "" {
		errs = append(errs, common.NewValidationError("new_password", common.ErrRequired))
	} else if err := r.NewPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("new_password", err))
	}

	// Check if current and new passwords are the same
	if r.CurrentPassword != "" && r.NewPassword != "" && r.CurrentPassword == r.NewPassword {
		errs = append(errs, common.NewValidationError("new_password", common.ErrNewPasswordSameAsCurrent))
	}

	return errs
}

// ============================================
// User Management (Filter Users)
// ============================================

type OrgRole string

const (
	OrgRoleInviteUsers OrgRole = "invite_users"
	OrgRoleManageUsers OrgRole = "manage_users"
)

type OrgUser struct {
	EmailAddress common.EmailAddress `json:"email_address"`
	Name         string              `json:"name"`
	Status       string              `json:"status"`
	CreatedAt    string              `json:"created_at"`
	Roles        []OrgRole           `json:"roles"`
}

type FilterOrgUsersRequest struct {
	Limit        *int32  `json:"limit,omitempty"`
	Cursor       *string `json:"cursor,omitempty"`
	FilterEmail  *string `json:"filter_email,omitempty"`
	FilterName   *string `json:"filter_name,omitempty"`
	FilterStatus *string `json:"filter_status,omitempty"`
}

func (r FilterOrgUsersRequest) Validate() []common.ValidationError {
	// Optional fields
	return nil
}

type FilterOrgUsersResponse struct {
	Items      []OrgUser `json:"items"`
	NextCursor string    `json:"next_cursor"`
}

// ===================================
// Language Management
// ===================================

type OrgSetLanguageRequest struct {
	Language common.LanguageCode `json:"language"`
}

func (r OrgSetLanguageRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Language == "" {
		errs = append(errs, common.NewValidationError("language", common.ErrRequired))
	} else if err := r.Language.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("language", err))
	}

	return errs
}
